src/storage/leveldb/address_index.ts

```diff
@@ -0,0 +1,341 @@
+    this.addressesIndexDB = db.sublevel(INDEX_PREFIX);
+    this.addressesMetaDB = db.sublevel<string, IAddressMetadataAsRecord>(ADDRESS_META_PREFIX, {valueEncoding: 'json'});
+    this.isValidated = false;
+  }
+
+  /**
+   * Close the database and the sublevel children.
+   * @returns {Promise<void>}
+   */
+  async close(): Promise<void> {
+    await this.addressesDB.db.close();
+  }
+
+  /**
+   * Check that the index version matches the expected version.
+   * @returns {Promise<void>}
+   */
+  async checkVersion(): Promise<void> {
+    const db = this.addressesDB.db;
+    try {
+      const dbVersion = await db.get('version');
+      if (this.indexVersion !== dbVersion) {
+        throw new Error(`Database version mismatch for ${this.constructor.name}: database version (${dbVersion}) expected version (${this.indexVersion})`);
+      }
+    } catch (err: unknown) {
+      if (errorCodeOrNull(err) === KEY_NOT_FOUND_CODE) {
+        // This is a new db, add version and return
+        await db.put('version', this.indexVersion);
+        return;
+      }
+
+      throw err;
+    }
+  }
+
+  /**
+   * Validate the index consistency along with the sublevel children.
+   * We check that all addresses in the addressesDB have an index in the addressesIndexDB.
+   * @returns {Promise<AddressIndexValidateResponse>} The first and last index in the database
+   */
+  async validate(): Promise<AddressIndexValidateResponse> {
+    // Clear metadata since we cannot guarantee the validity
+    await this.addressesMetaDB.clear();
+
+    const ret: AddressIndexValidateResponse = {firstIndex: Infinity, lastIndex: -1};
+    let size = 0;
+    // Iterate on all addresses and check that we have a corresponding index entry
+    for await (const [key, value] of this.addressesDB.iterator()) {
+      // increment size counter
+      size++;
+      if (key !== value.base58) {
+        throw new Error('Inconsistent database');
+      }
+      if (value.bip32AddressIndex < ret.firstIndex) {
+        ret.firstIndex = value.bip32AddressIndex;
+      }
+      if (value.bip32AddressIndex > ret.lastIndex) {
+        ret.lastIndex = value.bip32AddressIndex;
+      }
+
+      // check that we have an index
+      try {
+        const addressFromIndex = await this.addressesIndexDB.get(_index_key(value.bip32AddressIndex));
+        if (value.base58 !== addressFromIndex) {
+          throw new Error('Inconsistent database');
+        }
+      } catch(err: unknown) {
+        if (errorCodeOrNull(err) === KEY_NOT_FOUND_CODE) {
+          // Create if index is missing
+          await this.addressesIndexDB.put(_index_key(value.bip32AddressIndex), value.base58);
+          continue;
+        }
+        throw err;
+      }
+    }
+
+    // We just did a full address count, we can save the size and trust it
+    const buf = Buffer.alloc(4);
+    buf.writeUInt32BE(size);
+    await this.addressesDB.db.put('size', buf.toString('hex'));
```

Comment:
This code was added in 3 places and you have a method (`_index_key`) that would help a lot with it. I would change the name of the method to be more generic and use it.

Reply:
This has changed to a counter on memory